paxml/train.py

```diff
@@ -1114,40 +1120,48 @@ def _train_and_evaluate_common(
             'num_train_step (`%d`).', step_i, train_p.num_train_steps)
         break
 
-      # Get new model inputs
-      logging.debug('  Retrieving inputs.')
-      # TODO(hthu): Probably we should just internalize directly into step().
-      model_inputs = train_program.train_input.get_next_padded()
-      model_inputs = partitioner.preprocess_inputs(
-          train_program.train_input,
-          model_inputs,
-          train_program.partitioned_input_spec,
-      )
-      logging.debug('  Retrieved inputs.')
-
-      do_profile = train_p.profiler_capture_step is not None
-      if (do_profile and
-          step_i - initial_step == train_p.profiler_capture_step):
-        profiler.capture_async()
-
-      logging.debug('  Performing train_step().')
-      with jax.profiler.StepTraceAnnotation('train', step_num=step_i):
-        with py_utils.timeit() as train_period:
-          program_output = train_program.run_step(
-              partitioned_train_state,
-              train_prng_seed,
-              model_inputs,
-              train_program.train_unpadded_global_batch_size,
-          )
-          # TODO(hthu): Pass program output around instead of unpacking.
-          partitioned_train_state = program_output.state
-          loss = program_output.aux['loss']
-          weighted_scalars = program_output.aux['weighted_scalars']
-          per_example_out = program_output.aux['per_example_out']
-          summary_tensors = program_output.aux['summary_tensors']
+      loss_tracker = []
+      for ga_step in range(num_sub_batches):
+
+        # Get new model inputs
+        logging.debug('  Retrieving inputs.')
+        # TODO(hthu): Probably we should just internalize directly into step().
+        model_inputs = train_program.train_input.get_next_padded()
+        model_inputs = partitioner.preprocess_inputs(
+            train_program.train_input,
+            model_inputs,
+            train_program.partitioned_input_spec,
+        )
+        logging.debug('  Retrieved inputs.')
+
+        do_profile = train_p.profiler_capture_step is not None
+        if (do_profile and
+            step_i - initial_step == train_p.profiler_capture_step):
+          profiler.capture_async()
+
+        logging.debug('  Performing train_step().')
+        with jax.profiler.StepTraceAnnotation('train', step_num=step_i):
+          with py_utils.timeit() as train_period:
+            program_output = train_program.run_step(
+                partitioned_train_state,
+                train_prng_seed,
+                model_inputs,
+                train_program.train_unpadded_global_batch_size,
+            )
+            # TODO(hthu): Pass program output around instead of unpacking.
+            partitioned_train_state = program_output.state
+            loss = program_output.aux['loss']
```

Comment:
`loss` is asynchronously computed by JAX backend, see
https://jax.readthedocs.io/en/latest/async_dispatch.html#asynchronous-dispatch

when you ask for `loss` actual value, you will be blocking the main thread from dispatching the next GPU computation, so ideally the mean_loss_over_subbatches need to be async computed as well

the way to do that is avoid host computation (numpy or statistics will compute on CPU)

instead of loss_tracker=[], i recommend 

```
summed_loss = jnp.zeros((), dtype=jnp.float32)
for grad_accum_step in range(num_sub_batches):
  program_output = pjitted_train_step(...)
  loss = program_output.aux['loss']
  summed_loss = summed_loss + loss
loss = summed_loss / num_sub_batches
```

The add and divide will now happen on GPU as async dispatched computation

Reply:
ignore the last comment, `summed_loss = summed_loss + loss` should just work

there's no communications involved, so pjit=jit